File: Codes/Bingran Li/Tree/112. Path Sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean hasPathSum(TreeNode root, int sum) {
        
        if(root==null)
            return false;
        
        if(root.val==sum && root.left==null && root.right==null) // not consider root to LEAVE
            return true;
       
        // the value can be negative
            sum = sum - root.val;
            if(root.left!=null && hasPathSum(root.left,sum)) return true;
            if(root.right!=null && hasPathSum(root.right,sum)) return true;
        
        return false;
        
    }
}

File: Codes/Kan Zheng/C++/Tree/112. Path Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    //can't figure out how to handle [] 0
    // bool hasPathSum(TreeNode* root, int sum) {
    //     if(root==NULL&&sum!=0)
    //         return false;
    //     else if (root==NULL&&sum==0)
    //         return true;
    //     else if(hasPathSum(root->left,sum-root->val)||hasPathSum(root->right,sum-root->val))
    //         return true;
    // }
    bool hasPathSum(TreeNode* root, int sum) {
        if(root==NULL)
            return false;
        else if (root->left==NULL&&root->right==NULL){
            if(root->val==sum)
                return true;
            else
                return false;
        }
        else if (root->left==NULL&&root->right!=NULL)
            return hasPathSum(root->right, sum-root->val);
        else if (root->left!=NULL&&root->right==NULL)
            return hasPathSum(root->left, sum-root->val);
        else
            return hasPathSum(root->left,sum-root->val)||hasPathSum(root->right,sum-root->val);
    }
};